### ISSUE TITLE
Ignore files specified by .gitignore

### ISSUE THREAD

**Issue by rivertam** [junegunn/fzf.vim#121]
Would it be possible to have a command that won't return files in .gitignore (and/or other similar ignore configs)? I know Ag will do that, but it seems Ag searches through the files, not through the file names. :GitFiles is close, but if I have brand new files that have yet to be added to git (but are not in .gitignore), it doesn't detect them.

**Comment by junegunn** [comment]
See https://github.com/junegunn/fzf#respecting-gitignore-hgignore-and-svnignore
Note that you can define env vars in your vimrc as well

```vim
let $FZF_DEFAULT_COMMAND = 'ag -g ""'
```

**Comment by rivertam** [comment]
Thank you! I'd been specifying this in my zshrc, but apparently it wasn't propagating to vim for whatever reason.

**Comment by junegunn** [comment]
Hmm, it normally should.

**Comment by rivertam** [comment]
In case you're curious, I'm on Mac OSX using iTerm 2 and oh-my-zsh with Neovim. `echo $FZF_DEFAULT_COMMAND` yields `ag -g ""`

**Comment by junegunn** [comment]
`:echo $FZF_DEFAULT_COMMAND` on Vim should give the same result.

**Comment by rivertam** [comment]
Hmm. Maybe the issue was that I hadn't restarted vim since I changed the environment variable? I'm not sure, but it seems to actually be working that way now as well. Sorry for the inconvenience and thanks for the help. Big fan. =)

**Comment by rivertam** [comment]
It seems like this is a neovim issue or an integration issue, but vim is back to not recognizing the variable (`:echo $FZF_DEFAULT_COMMAND` prints nothing). I'm going to put the let statement in my init.vim so this won't continue to be an issue, but I suppose it's worth noting.

**Comment by balsagoth** [comment]
I'm using neovim too, and `:echo $FZF_DEFAULT_COMMAND` prints notthing

**Comment by rivertam** [comment]
@balsagoth Just to be clear, what does `echo $FZF_DEFAULT_COMMAND` do when you're in bash/zsh?

**Comment by balsagoth** [comment]
@rivertam same as you. I'm using zsh
```
 > echo $FZF_DEFAULT_COMMAND
ag -g ""
```

**Comment by junegunn** [comment]
@balsagoth Did you export the variable? `export FZF_DEFAULT_COMMAND`

**Comment by noscripter** [comment]
https://github.com/junegunn/fzf/issues/383#issuecomment-278013803

Use `:GFiles` plz.

**Comment by Lesik** [comment]
@rivertam @noscripter Keep in mind that `:GFiles` is using `git ls-files`
https://github.com/junegunn/fzf.vim/blob/9b50e2fd5f7beab1cac8269cfc386c86d266df38/autoload/fzf/vim.vim#L536-L547
which means that it not only ignores files in `.gitignore`, but also files not added to git. So if you freshly create a file, you won't find it in `:GFiles`.

On the other hand, the suggested solution to overwrite `$FZF_DEFAULT_COMMAND` to something like `rg --files` hides `.gitignore` files but shows freshly created files, like OP asked.

**Comment by mreishus** [comment]
for ripgrep users:

```viml
" Tell FZF to use RG - so we can skip .gitignore files even if not using
" :GitFiles search
let $FZF_DEFAULT_COMMAND = 'rg --files --hidden'
" If you want gitignored files:
"let $FZF_DEFAULT_COMMAND = 'rg --files --no-ignore-vcs --hidden'
```

**Comment by hvladev** [comment]
**tl;dr**

```
:GFiles --cached --others --exclude-standard
```

---

**Why this works?**

Options passed to `GFiles` will be then passed to `git ls-files`.

https://github.com/junegunn/fzf.vim/blob/dc4c4c22715c060a2babd5a5187004bdecbcdea7/autoload/fzf/vim.vim#L536-L547

I'm using the following options to include git files, untracked files and to exclude git ignored files:

```
:GFiles --cached --others --exclude-standard
```

Here's what these options do ([`man git-ls-files`](https://git-scm.com/docs/git-ls-files)):
```
OPTIONS
       -c, --cached
           Show cached files in the output (default)
...
       -o, --others
           Show other (i.e. untracked) files in the output
...
       --exclude-standard
           Add the standard Git exclusions: .git/info/exclude, .gitignore in each directory, and the user's global exclusion file.
```

Finally, you can map it to whatever you want:

```vim script
nnoremap <silent> <leader>j :GFiles --cached --others --exclude-standard<cr>
```

**Comment by ayushxx7** [comment]
How to do the opposite? Show .gitignore files? But not always.
map1 (say `Ctrl-f`) will point to search with files hidden,
while map2 (say `Ctrl-h`) will search all files (including ignored)

Currently, I have to comment this line out:
```
" let $FZF_DEFAULT_COMMAND="rg --files --hidden"
```

**Comment by lyonsun** [comment]
`ripgrep` might be the easiest way to go now that it by default respects the `.gitignore` file.

> By default, ripgrep will respect gitignore rules and automatically skip hidden files/directories and binary files.
